### PR TITLE
configmap item with order

### PR DESCRIPTION
**Pull request by charles-chenzz**
this PR tend to solve #4292 
since #5210 remain stagnant, so I took it up.
Also, I borrow some of the code in #5210

### CODE REVIEW

**Comment by charles-chenzz** [comment]
the lint checks failed, but wasn't about the mismatch of args. have no clues why it would failed. would defer till I fix the Test Linux and Test MacOS CI jobs.


```
Error: api/resmap/factory.go:105:2: Consider pre-allocating `resources` (prealloc)
	var (
	^
Error: api/internal/generators/configmap.go:41:20: error returned from external package is unwrapped: sig: func (*sigs.k8s.io/kustomize/kyaml/yaml.RNode).LoadMapIntoConfigMapData(m map[string]string) error (wrapcheck)
		return nil, nil, err
		                 ^
Error: api/internal/generators/secret.go:47:20: error returned from external package is unwrapped: sig: func (*sigs.k8s.io/kustomize/kyaml/yaml.RNode).Pipe(functions ...sigs.k8s.io/kustomize/kyaml/yaml.Filter) (*sigs.k8s.io/kustomize/kyaml/yaml.RNode, error) (wrapcheck)
		return nil, nil, err
		                 ^
Error: api/internal/generators/secret.go:54:20: error returned from external package is unwrapped: sig: func (*sigs.k8s.io/kustomize/kyaml/yaml.RNode).LoadMapIntoSecretData(m map[string]string) error (wrapcheck)
		return nil, nil, err
		                 ^
Error: api/internal/generators/utils.go:50:21: error returned from interface method should be wrapped: sig: func (sigs.k8s.io/kustomize/api/ifc.Validator).ErrIfInvalidKey(string) error (wrapcheck)
			return nil, nil, err
			                 ^
Error: api/resource/factory.go:281:20: error returned from external package is unwrapped: sig: func sigs.k8s.io/kustomize/api/internal/generators.MakeConfigMap(ldr sigs.k8s.io/kustomize/api/ifc.KvLoader, args *sigs.k8s.io/kustomize/api/types.ConfigMapArgs) (rn *sigs.k8s.io/kustomize/kyaml/yaml.RNode, orderKeys []string, err error) (wrapcheck)
		return nil, nil, err
		                 ^
Error: api/resource/factory.go:290:20: error returned from external package is unwrapped: sig: func sigs.k8s.io/kustomize/api/internal/generators.MakeSecret(ldr sigs.k8s.io/kustomize/api/ifc.KvLoader, args *sigs.k8s.io/kustomize/api/types.SecretArgs) (rn *sigs.k8s.io/kustomize/kyaml/yaml.RNode, orderKeys []string, err error) (wrapcheck)
		return nil, nil, err
		                 ^
Error: api/resmap/factory.go:96:20: error returned from external package is unwrapped: sig: func (*sigs.k8s.io/kustomize/api/resource.Factory).MakeConfigMap(kvLdr sigs.k8s.io/kustomize/api/ifc.KvLoader, args *sigs.k8s.io/kustomize/api/types.ConfigMapArgs) (*sigs.k8s.io/kustomize/api/resource.Resource, []string, error) (wrapcheck)
		return nil, nil, err
		                 ^
Error: api/resmap/factory.go:127:20: error returned from external package is unwrapped: sig: func (*sigs.k8s.io/kustomize/api/resource.Factory).MakeSecret(kvLdr sigs.k8s.io/kustomize/api/ifc.KvLoader, args *sigs.k8s.io/kustomize/api/types.SecretArgs) (*sigs.k8s.io/kustomize/api/resource.Resource, []string, error) (wrapcheck)
		return nil, nil, err
		                 ^
make[1]: *** [../Makefile-modules.mk:18: lint] Error 1
make[1]: Leaving directory '/home/runner/work/kustomize/kustomize/api'
make: *** [Makefile:127: lint] Error 2
Error: Process completed with exit code 2.
```

**Comment by natasha41575** [comment]
` error returned from external package is unwrapped` means you need to wrap the error, like

```
return nil, nil, fmt.Errorf("failed to xxxx: %w", err)
```

**Comment by charles-chenzz** [comment]
I convert to ready for review for gathering some helps and reviews

**Review by stormqueen1990 — COMMENTED** [review]
Hi there, @charles-chenzz! 👋

I left some suggestions. I can also reproduce the error you're seeing about mismatched arguments locally after changing the signature of `Generate` in `converter_test.go`, so I'm assuming the change in the method signature might require more steps to pass tests. Perhaps @natasha41575 or @annasong20 can give us some help to understand the issue?

File: api/internal/generators/utils.go [hunk 1]
```diff
@@ -37,24 +37,27 @@ kind: %s
...
-			return nil, err
+			return nil, nil, fmt.Errorf("failed to validate key, error:%w", err)
 		}
 		if _, ok := knownKeys[p.Key]; ok {
-			return nil, errors.Errorf(
+			return nil, nil, errors.Errorf(
 				"configmap %s illegally repeats the key `%s`", name, p.Key)
 		}
 		knownKeys[p.Key] = p.Value
+		orderKeys = append(orderKeys, p.Key)
```
[the remark this answers is not in the document]
Comment (charles-chenzz) [reply api/internal/generators/utils.go:57]: was trying to using a slice of string to preserve the order when for range the pairs, then return it. I was based on the issue and the former PR as my context about this feature. but what I commit in this PR might have details that wrong or misunderstood.

[the remark this answers is not in the document]
Comment (stormqueen1990) [reply api/internal/generators/utils.go:57]: I think there might be some extra steps needed. Just adding the keys to a slice will likely not be enough to guarantee a predictable order. Another thing I noticed while investigating this code is that the output from this method is discarded, so what ends up being written to the file is exactly the user input: https://github.com/kubernetes-sigs/kustomize/blob/7c36ed21b3587e7124326743a888db5374e86d93/kustomize/commands/edit/add/configmap.go#L114-L116

File: api/resmap/resmap.go [hunk 2]
```diff
@@ -30,7 +30,7 @@ type TransformerWithProperties struct {
 
 // A Generator creates an instance of ResMap.
 type Generator interface {
-	Generate() (ResMap, error)
+	Generate() (ResMap, []string, error)
```
Comment (natasha41575) [inline api/resmap/resmap.go:33]: So unfortunately, this is a public interface and we can't change the signature (this is how users built legacy style extensions and we can't break that just yet). 

I think it also isn't quite sensical for this Generator to return an array of keys just for a ConfigMap. That type of data is specific to the ConfigMap generator and is not generalizable to all generators, so should instead be attached to the ConfigMapGenerator and methods - and only to things that are specific to the ConfigMapGenerator.

Comment (charles-chenzz) [reply api/resmap/resmap.go:33]: I would think about other solution to see if we change limit the changes only to configmapgenerator

**Pull request closed without merging**